File: data-structures/python/src/stack.py
class Stack:
    def __init__(self):
        self._data = []

    def push(self, value):
        self._data.append(value)

    def pop(self):
        if not self._data:
            raise IndexError("pop from empty stack")
        return self._data.pop()

    def top(self):
        if not self._data:
            raise IndexError("top from empty stack")
        return self._data[-1]

    def size(self):
        return len(self._data)

    def is_empty(self):
        return len(self._data) == 0

    def clear(self):
        self._data.clear()

    def copy(self):
        clone = Stack()
        clone._data = self._data.copy()
        return clone

    def __len__(self):
        return len(self._data)

    def __bool__(self):
        return len(self._data) > 0

File: data-structures/go/src/stack/stack.go
package stack

import "errors"

var ErrEmptyStack = errors.New("Stack: empty stack")

type Stack[T any] struct {
	data []T
}

func New[T any]() *Stack[T] {
	return &Stack[T]{}
}

func (s *Stack[T]) Push(value T) {
	s.data = append(s.data, value)
}

func (s *Stack[T]) Pop() (T, error) {
	var zero T
	if len(s.data) == 0 {
		return zero, ErrEmptyStack
	}
	idx := len(s.data) - 1
	value := s.data[idx]
	s.data[idx] = zero
	s.data = s.data[:idx]
	return value, nil
}

func (s *Stack[T]) Top() (T, error) {
	var zero T
	if len(s.data) == 0 {
		return zero, ErrEmptyStack
	}
	return s.data[len(s.data)-1], nil
}

func (s *Stack[T]) Size() int {
	return len(s.data)
}

func (s *Stack[T]) IsEmpty() bool {
	return len(s.data) == 0
}

func (s *Stack[T]) Clear() {
	var zero T
	for i := range s.data {
		s.data[i] = zero
	}
	s.data = s.data[:0]
}

func (s *Stack[T]) Clone() *Stack[T] {
	clone := &Stack[T]{
		data: make([]T, len(s.data)),
	}
	copy(clone.data, s.data)
	return clone
}
